e2e_tests/e2e_test.go

```diff
@@ -0,0 +1,794 @@
+		retryUpgrade.Install = true
+		_, err = retryUpgrade.Run("sock", chart, values)
+		require.NoError(t, err)
+	}
+
+	waitForAllDeploymentsToStart(t, clientset)
+
+	deployments := clientset.AppsV1().Deployments("default")
+
+	decode := scheme.Codecs.UniversalDeserializer().Decode
+	stream, err := os.ReadFile(filepath.Join("testdata", "nodejs", "deployment.yaml"))
+	require.NoError(t, err)
+	deployment, _, err := decode(stream, nil, nil)
+	require.NoError(t, err)
+	_, err = deployments.Create(context.Background(), deployment.(*appsv1.Deployment), metav1.CreateOptions{})
+	if err != nil {
+		_, err2 := deployments.Update(context.Background(), deployment.(*appsv1.Deployment), metav1.UpdateOptions{})
+		assert.NoError(t, err2)
+		if err2 != nil {
+			require.NoError(t, err)
+		}
+	}
+	jobstream, err := os.ReadFile(filepath.Join("testdata", "test_jobs.yaml"))
+	require.NoError(t, err)
+	var namespaces []*corev1.Namespace
+	var jobs []*batchv1.Job
+	for _, resourceYAML := range strings.Split(string(jobstream), "---") {
+		if len(resourceYAML) == 0 {
+			continue
+		}
+
+		obj, groupVersionKind, err := decode(
+			[]byte(resourceYAML),
+			nil,
+			nil)
+		require.NoError(t, err)
+		if groupVersionKind.Group == "" &&
+			groupVersionKind.Version == "v1" &&
+			groupVersionKind.Kind == "Namespace" {
+			nm := obj.(*corev1.Namespace)
+			namespaces = append(namespaces, nm)
+			nms := clientset.CoreV1().Namespaces()
+			_, err := nms.Create(context.Background(), nm, metav1.CreateOptions{})
+			require.NoError(t, err)
+			t.Logf("Deployed namespace %s", nm.Name)
+		}
+
+		waitForAllNamespacesToBeCreated(t, clientset)
+
+		if groupVersionKind.Group == "batch" &&
+			groupVersionKind.Version == "v1" &&
+			groupVersionKind.Kind == "Job" {
+			job := obj.(*batchv1.Job)
+			jobs = append(jobs, job)
+			jobClient := clientset.BatchV1().Jobs(job.Namespace)
+			_, err := jobClient.Create(context.Background(), job, metav1.CreateOptions{})
+			require.NoError(t, err)
+			t.Logf("Deployed job %s", job.Name)
+		}
+	}
+
+	waitForAllDeploymentsToStart(t, clientset)
+
+	t.Cleanup(func() {
+		if os.Getenv("SKIP_TEARDOWN") == "true" {
+			t.Log("Skipping teardown as SKIP_TEARDOWN is set to true")
+			return
+		}
+		waitTime := int64(0)
+		_ = deployments.Delete(context.Background(), "nodejs-test", metav1.DeleteOptions{
+			GracePeriodSeconds: &waitTime,
+		})
+		for _, job := range jobs {
+			jobClient := clientset.BatchV1().Jobs(job.Namespace)
+			_ = jobClient.Delete(context.Background(), job.Name, metav1.DeleteOptions{
+				GracePeriodSeconds: &waitTime,
+			})
+		}
+		for _, nm := range namespaces {
+			nmClient := clientset.CoreV1().Namespaces()
+			_ = nmClient.Delete(context.Background(), nm.Name, metav1.DeleteOptions{
+				GracePeriodSeconds: &waitTime,
+			})
+		}
+		uninstall := action.NewUninstall(actionConfig)
+		uninstall.IgnoreNotFound = true
+		uninstall.Wait = true
+		_, _ = uninstall.Run("sock")
+	})
+}
+
+func Test_E2E(t *testing.T) {
```

Comment:
This test is doing a lot. Can it be broken up in to individual cases (maybe using a testify suite) instead of requiring an all or nothing execution?

Reply:
this test runs subtests with t.Run - sorry if I missed something, but is that not how a test suite would be done in Golang? It is using a setupOnce function to set up the environment so it is shared across tests so they quickly spin up.

Reply:
And the testing.T pointer passed around is used in setupOnce to register cleanups with t.Cleanup to destroy the helm chart and apps deployed after run, hence the need to organize with sub tests.